cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/sourcecontrol/ApplicationManagerTest.java

```diff
@@ -0,0 +1,194 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.sourcecontrol;
+
+import io.cdap.cdap.AllProgramsApp;
+import io.cdap.cdap.api.artifact.ArtifactSummary;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.id.Id;
+import io.cdap.cdap.internal.app.services.http.AppFabricTestBase;
+import io.cdap.cdap.proto.ApplicationDetail;
+import io.cdap.cdap.proto.app.AppVersion;
+import io.cdap.cdap.proto.app.UpdateMultiSourceControlMetaReqeust;
+import io.cdap.cdap.proto.app.UpdateSourceControlMetaRequest;
+import io.cdap.cdap.proto.artifact.AppRequest;
+import io.cdap.cdap.proto.id.ApplicationId;
+import io.cdap.cdap.proto.id.ApplicationReference;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.sourcecontrol.operationrunner.PullAppResponse;
+import io.cdap.common.http.HttpResponse;
+import java.util.Arrays;
+import java.util.Collection;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+/**
+ * Tests for {@link RemoteApplicationManager} and {@link LocalApplicationManager}
+ */
+@RunWith(Parameterized.class)
+public class ApplicationManagerTest extends AppFabricTestBase {
+
+  private enum ApplicationManagerType {
+    LOCAL,
+    REMOTE,
+  }
+
+  private final ApplicationManagerType managerType;
+
+  private static final String namespace = TEST_NAMESPACE1;
+  private static final Id.Artifact artifact = Id.Artifact.from(new Id.Namespace(namespace),
+      AllProgramsApp.class.getSimpleName(), "1.0.0-SNAPSHOT");
+  private static final AppRequest<?> request = new AppRequest<>(
+      ArtifactSummary.from(artifact.toArtifactId()));
+
+  public ApplicationManagerTest(ApplicationManagerType type) {
+    this.managerType = type;
+  }
+
+  @Parameterized.Parameters
+  public static Collection<Object[]> parameters() {
+    return Arrays.asList(new Object[][]{
+        {ApplicationManagerType.LOCAL},
+        {ApplicationManagerType.REMOTE},
+    });
+  }
+
+  @SuppressWarnings("checkstyle:MissingSwitchDefault")
+  private ApplicationManager getApplicationManager(ApplicationManagerType type) {
+    if (type == ApplicationManagerType.LOCAL) {
+      return AppFabricTestBase.getInjector().getInstance(LocalApplicationManager.class);
+    }
+    return AppFabricTestBase.getInjector().getInstance(RemoteApplicationManager.class);
+  }
+
+
+  @Before
+  public void setUp() throws Exception {
+    HttpResponse response = addAppArtifact(artifact, AllProgramsApp.class);
+    Assert.assertEquals(200, response.getResponseCode());
+  }
+
+  @Test
+  public void testDeployApp() throws Exception {
```

Comment:
No test for failure cases as the functions are just calling APIs/service methods directly and throwing the same exception. So the failure cases would be covered by underlying API/method unit tests